### PR TITLE
feat: combine capabilities

### PR DESCRIPTION
**Pull request by alvin-reyes**
# Combine Manual and Auto-Detected Capabilities in Telemetry Reports

## Resolves 
https://github.com/masa-finance/tee-indexer/issues/235

## Problem
The current implementation prioritizes manual capability configuration over automatic detection. When `CAPABILITIES="all"` is set in the environment, the telemetry report only shows `"capability": "all"`, completely ignoring auto-detected capabilities. This limits transparency and makes it difficult to understand the worker's actual capabilities.

## Solution
This PR modifies the TEE Worker to combine manual and auto-detected capabilities in telemetry reports, ensuring complete visibility of all capabilities the worker can provide.

### 1. New Capability Detection Module
- Added `internal/capabilities/detector.go` with:
  - `DetectCapabilities()`: Auto-detects available capabilities based on credentials
  - `MergeCapabilities()`: Combines manual and auto-detected capabilities without duplication
  - Detects Twitter capabilities based on accounts/API keys
  - Always includes base capabilities (web-scraper, telemetry, tiktok-transcription)

### 2. Updated Stats Collector
- Modified `internal/jobs/stats/stats.go` to:
  - Use the new capability detector
  - Merge manual capabilities from environment with auto-detected ones
  - Log combined capabilities for transparency

### 3. Enhanced Documentation
- Updated README.md to clarify:
  - How capabilities are detected and reported
  - That manual and auto-detected capabilities are combined
  - Examples of expected behavior

### 4. Comprehensive Tests
- Added `internal/capabilities/detector_test.go` with tests for:
  - Various credential configurations
  - Capability merging logic
  - Edge cases and validation

### Before
```json
// With CAPABILITIES="all" and Twitter credentials
{
  "reported_capabilities": ["all"]
}
```

### After
```json
// With CAPABILITIES="all" and Twitter credentials
{
  "reported_capabilities": ["all", "searchbyquery", "getbyid", "gettweets", "web-scraper", "telemetry", ...]
}
```

## Testing
- ✅ All unit tests pass
- ✅ Build completes successfully
- ✅ Backward compatibility maintained
- ✅ Manual capability configuration still works as before

## Impact
- **No breaking changes** - Existing configurations continue to work
- **Enhanced transparency** - Full visibility of worker capabilities
- **Better resource allocation** - MASA ecosystem can make informed decisions based on actual capabilities

## Related
- Addresses feedback from PR #81 about capability precedence
- Critical for accurate worker evaluation in the MASA ecosystem

## Verification
```
masa-tee-worker-1  | time="2025-06-24T15:22:49Z" level=info msg="Creating new TwitterXClient with API key"
masa-tee-worker-1  | time="2025-06-24T15:22:49Z" level=info msg="TwitterXClient instantiated successfully using base URL: https://api.x.com/2"
masa-tee-worker-1  | time="2025-06-24T15:22:49Z" level=info msg="GET request to: https://api.x.com/2/tweets/search/all?query=from:twitterdev&max_results=10"
masa-tee-worker-1  | time="2025-06-24T15:22:49Z" level=info msg="TikTokTranscriber initialized with configuration" api_origin="https://submagic-free-tools.fly.dev" api_referer="https://submagic-free-tools.fly.dev/tiktok-transcription" api_user_agent="Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/135.0.0.0 Mobile Safari/537.36" default_language=eng-US transcription_endpoint="https://submagic-free-tools.fly.dev/api/tiktok-transcription"
masa-tee-worker-1  | time="2025-06-24T15:22:49Z" level=info msg="Initialized job worker for: web-scraper"
masa-tee-worker-1  | time="2025-06-24T15:22:49Z" level=info msg="Initialized job worker for: twitter-scraper"
masa-tee-worker-1  | time="2025-06-24T15:22:49Z" level=info msg="Initialized job worker for: twitter-credential-scraper"
masa-tee-worker-1  | time="2025-06-24T15:22:49Z" level=info msg="Initialized job worker for: twitter-api-scraper"
masa-tee-worker-1  | time="2025-06-24T15:22:49Z" level=info msg="Initialized job worker for: telemetry"
masa-tee-worker-1  | time="2025-06-24T15:22:49Z" level=info msg="Initialized job worker for: tiktok-transcription"
masa-tee-worker-1  | time="2025-06-24T15:22:49Z" level=info msg="Job workers setup completed."
masa-tee-worker-1  | time="2025-06-24T15:22:49Z" level=info msg="JobServer initialization complete."
**masa-tee-worker-1  | time="2025-06-24T15:22:49Z" level=info msg="Updated capabilities with full detection (manual + worker-reported): [searchbyquery getbyid getprofilebyid searchbyfullarchive telemetry tiktok-transcription web-scraper]"
masa-tee-worker-1  | {"time":"2025-06-**24T15:22:49.306714326Z","level":"INFO","prefix":"echo","file":"start.go","line":"87","message":"Enabling profiling control endpoints"}
masa-tee-worker-1  | {"time":"2025-06-24T15:22:49.306714326Z","level":"INFO","prefix":"echo","file":"start.go","line":"125","message":"Starting server on :8083"}
masa-tee-worker-1  | 
masa-tee-worker-1  |    ____    __

```

### CODE REVIEW

File: internal/capabilities/detector.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+		// For now, we'll check if accounts are in the correct format (username:password)
+		validAccounts := 0
+		for _, acc := range accounts {
+			if strings.Contains(acc, ":") {
+				validAccounts++
+			}
+		}
+		if validAccounts > 0 {
+			// Credential-based capabilities
+			capabilities = append(capabilities,
```
[the remark this answers is not in the document]
Comment (mcamou) [reply internal/capabilities/detector.go:48]: yeah I think it's better to have it per-scraper. That way each new scraper is self-contained, otherwise we would have to edit this detector object each time we add a new scraper (case in point: LinkedIn, which is a separate repo).

File: internal/capabilities/detector.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+// MergeCapabilities combines manual and auto-detected capabilities
+func MergeCapabilities(manual string, detected []string) []string {
+	// Parse manual capabilities
+	var manualCaps []string
+	if manual != "" {
+		if strings.Contains(manual, ",") {
+			manualCaps = strings.Split(manual, ",")
+		} else {
+			manualCaps = []string{manual}
+		}
```
Comment (mcamou) [inline internal/capabilities/detector.go:69]: `strings.Split` will return a single-element slice if the delimiter is not found

```suggestion
		manualCaps = strings.Split(manual, ",")
```

File: internal/capabilities/detector.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		// Trim whitespace
+		for i := range manualCaps {
+			manualCaps[i] = strings.TrimSpace(manualCaps[i])
+		}
+	}
+
+	// Use a map to deduplicate
+	capMap := make(map[string]bool)
```
Comment (mcamou) [inline internal/capabilities/detector.go:77]: [nitpick] we could use a `map[string]struct{}` which saves a bit of memory.

File: internal/capabilities/detector.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	// Fallback to basic detection if no JobServer is available
+	// This maintains backward compatibility and is used during initialization
+	
+	// Always available capabilities
+	detected = append(detected, "web-scraper", "telemetry", "tiktok-transcription")
+	
+	// Check for Twitter capabilities based on credentials
+	if accounts, ok := jc["twitter_accounts"].([]string); ok && len(accounts) > 0 {
+		// Basic Twitter capabilities when accounts are available
+		detected = append(detected, "searchbyquery", "getbyid", "getprofilebyid")
```
Comment (mcamou) [inline internal/capabilities/detector.go:37]: as a TODO item for the future: instead of using just strings, let's use stronger typing. As in:

```
type Capability string

const (
  SearchByQuery Capability = "searchbyquery"
  GetById Capability = "getbyid"
  // and so on
)
```

If we then add them to `tee-types` we can e.g. have a `capability.Match(jt JobType)` method which takes into account wildcards. This needs a bit more design, is it `capability.Match(jt JobType)` or `jobType.Match(capabilities ...Capability)`. I think the second is more ergonomic.

This makes APIs more self-documenting, as well as preventing misspells.

And let's keep this pattern in mind for the future, please.

File: internal/capabilities/detector_test.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+import (
+	"reflect"
+	"sort"
+	"testing"
+
+	"github.com/masa-finance/tee-worker/api/types"
+)
+
+// MockJobServer implements JobServerInterface for testing
```
Comment (mcamou) [inline internal/capabilities/detector_test.go:11]: Very nice testing, love it!

File: internal/capabilities/detector.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	}
+	
+	if apiKeys, ok := jc["twitter_api_keys"].([]string); ok && len(apiKeys) > 0 {
+		// Basic API capabilities
+		if !contains(detected, "searchbyquery") {
```
Comment (mcamou) [inline internal/capabilities/detector.go:42]:
```suggestion
		if !slices.Contains(detected, "searchbyquery") {
```

File: internal/capabilities/detector.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+func contains(slice []string, item string) bool {
+	for _, s := range slice {
+		if s == item {
+			return true
+		}
+	}
+	return false
+}
```
Comment (mcamou) [inline internal/capabilities/detector.go:57]: Please replace with `slices.Contains`
```suggestion

```

**Review by mcamou — COMMENTED** [review]
Very nice refactoring. Looking good, just a couple more comments and a TODO item.

File: internal/capabilities/detector.go (hunk 3 above)
Comment (mcamou) [reply internal/capabilities/detector.go:64]: sounds good
